core/core.go

```diff
@@ -0,0 +1,125 @@
+package core
+
+import (
+	"errors"
+	"io"
+	"time"
+)
+
+// Errors
+var (
+	ErrProcessNotFinished     = errors.New("Error: Process not finished")
+	ErrProcessNotStarted      = errors.New("Error: process not started yet")
+	ErrProcessAlreadyFinished = errors.New("Error: process already finished")
+	ErrProcessAlreadyStarted  = errors.New("Error: process already started")
+)
+
+// AppBus signals
+var (
+	SignalBuildComplete = `/build/complete/(?P<token>\w+)$`
+	SignalBuildStarted  = `/build/started/(?P<token>\w+)$`
+)
+
+type (
+	// App is defined by $NGBUILD_WORKSPACE/apps/$appname/config.yaml
+	// Apps will define different builds, for different projects or different kinds of builds or whatever
+	// Config has $NGBUILD_WORKSPACE/config.yaml as a parent and then applies
+	// $NGBUILD_WORKSPACE/apps/$appname/config.yaml onto it
+	// Everything App should be thread safe as it will be called from many goroutines
+	App interface {
+		Name() string
+		Config(namespace string, conf interface{}) error
+
+		// SendEvent is a dispatcher, it will send this string across all the apps integrations and also Core
+		SendEvent(event string)
+
+		// Listen will provide a channel to select on for a given regular expression
+		// returned map is the captured groups and values
+		Listen(event string) <-chan map[string]string
+
+		// NewBuild will be used by github and the like to create new builds for this app whenever they deem so
+		NewBuild(group string, config *BuildConfig) (token string, err error)
+
+		GetBuild(token string) (Build, error)
+	}
+
+	// BuildConfig describes a build, heavily in favour of github/git at the moment
+	BuildConfig struct {
```

Comment:
Would be great to have:

- `Name string` - so the build could have a name i.e. the Pull Request title, provided by the integration
- `URL string` - for the link to the PR
- `metadata map[string]string` - so integrations could save some state when doing interactive stuff like buttons etc. Otherwise each integration will have it's own `token`->`state` map which it won't know when to clean up etc. (or add a signal for cleanup if don't like the metadata idea)